File: problems/chapter16/miiitomi/002.cpp
// https://onlinejudge.u-aizu.ac.jp/problems/2872

#include <iostream>
#include <vector>
using namespace std;
typedef long long ll;
const ll INF = (1LL << 60);

struct Edge {
    int rev, to, dis, cost;
    Edge(int r, int t, int d, int c) {
        rev = r;
        to = t;
        dis = d;
        cost = c;
    }
};

void WarshallFloyd(vector<vector<Edge>> &G, vector<vector<vector<ll>>> &dp) {
    int N = (int)G.size();
    dp.assign(N+1, vector<vector<ll>>(N, vector<ll>(N, INF)));
    for (int i = 0; i < N; i++) {
        dp[0][i][i] = 0;
        for (auto e : G[i]) {
            dp[0][i][e.to] = e.dis;
        }
    }

    for (int i = 1; i <= N; i++) {
        for (int j = 0; j < N; j++) {
            for (int k = 0; k < N; k++) {
                dp[i][j][k] = min(dp[i-1][j][k], dp[i-1][j][i-1] + dp[i-1][i-1][k]);
            }
        }
    }
}

int rec(vector<vector<Edge>> &H, vector<bool> &seen, int v, int c, int t) {
    if (seen[v]) return 0;
    seen[v] = true;
    if (v == t) return c;

    for (auto &e : H[v]) {
        if (e.cost == 0) continue;

        int x = rec(H, seen, e.to, min(c, e.cost), t);
        if (x > 0) {
            e.cost -= x;
            H[e.to][e.rev].cost += x;
            return x;
        }
    }
    return 0;
}

int FordFulkerson(vector<vector<Edge>> &H, int s, int t) {
    int ans = 0;
    while (true) {
        vector<bool> seen((int)H.size(), false);
        int x = rec(H, seen, s, 100, t);
        ans += x;
        if (x == 0) break;
    }
    return ans;
}

int main() {
    int N, M, s, t;
    cin >> N >> M >> s >> t;
    s--;
    t--;
    vector<vector<Edge>> G(N);
    for (int i = 0; i < M; i++) {
        int u, v, d, c;
        cin >> u >> v >> d >> c;
        u--;
        v--;
        G[u].push_back(Edge(0, v, d, c));
    }

    vector<vector<vector<ll>>> dp;
    WarshallFloyd(G, dp);

    vector<vector<Edge>> H(N);
    for (int i = 0; i < N; i++) {
        for (auto e : G[i]) {
            if (dp[N][s][t] == dp[N][s][i] + e.dis + dp[N][e.to][t]) {
                H[i].push_back(Edge((int)H[e.to].size(), e.to, e.dis, e.cost));
                H[e.to].push_back(Edge((int)H[i].size()-1, i, e.dis, 0));
            }
        }
    }

    cout << FordFulkerson(H, s, t) << endl;
}

File: problems/chapter16/mshibatatt/002.py
# https://onlinejudge.u-aizu.ac.jp/beta/room.html#RitsCamp18Day3/problems/F
import sys
sys.setrecursionlimit(10000)

def floyd_warshall(G):
    N = len(G)
    dp = [[float('Inf')] * N for _ in range(N)]
    # init condition
    for i in range(N):
        dp[i][i] = 0
    for i, e in enumerate(G):
        for nxt in e:
            dp[i][nxt[0]] = nxt[1]
    
    for k in range(N):
        for i in range(N):
            for j in range(N):
                dp[i][j] = min(dp[i][j], dp[i][k] + dp[k][j])
    
    return dp

class FordFulkerson:
    def __init__(self, N):
        self.N = N
        self.G = [[] for i in range(N)]
 
    def add_edge(self, fr, to, cap):
        forward = [to, cap, None]
        forward[2] = backward = [fr, 0, forward]
        self.G[fr].append(forward)
        self.G[to].append(backward)
 
    def add_multi_edge(self, v1, v2, cap1, cap2):
        edge1 = [v2, cap1, None]
        edge1[2] = edge2 = [v1, cap2, edge1]
        self.G[v1].append(edge1)
        self.G[v2].append(edge2)
 
    def dfs(self, v, t, f):
        if v == t:
            return f
        seen = self.seen
        seen[v] = True
        for e in self.G[v]:
            w, cap, rev = e
            if cap and not seen[w]:
                d = self.dfs(w, t, min(f, cap))
                if d:
                    e[1] -= d
                    rev[1] += d
                    return d
        return 0
 
    def flow(self, s, t):
        flow = 0
        f = INF = float('inf')
        N = self.N
        while f:
            self.seen = [False]*N
            f = self.dfs(s, t, INF)
            flow += f
        return flow

def main():
    N, M, s, t = map(int, input().split())
    s -= 1
    t -= 1
    G = [[] for _ in range(N)]
    for _ in range(M):
        u, v, d, c = map(int, input().split())
        u -= 1
        v -= 1
        G[u].append([v, d, c])

    # find DAG including paths that are necessary for shortest path
    dist = floyd_warshall(G)
    shortest_path_length = dist[s][t]
    DAG = FordFulkerson(N)
    for i, e in enumerate(G):
        for v, d, c in e:
            if dist[s][i] + d + dist[v][t] == shortest_path_length:
                DAG.add_edge(i, v, c)

    # find minimum cut
    output = DAG.flow(s, t)
    print(output)

    
if __name__=='__main__':
    main()
